### PR TITLE
Train single class models

### PR DESCRIPTION
**Pull request by roshankern**
This PR is ready for review!

In this PR, single-class models are trained with the `train` subset of labeled Mitocheck data. For each of the 15 phenotypic classes, a model was trained to determine whether or not a cell's features belong to a certain phenotype. These models with later be evaluated and compared to the multi-class models to determine which method of classification is better.

### CODE REVIEW

**Comment by roshankern** [comment]
For the sake of this PR review, all 96 models have been moved to another folder and will be added to this repo after the PR has been reviewed.

File: 2.train_model/README.md [hunk 1]
```diff
@@ -2,7 +2,10 @@
 
 In this module, we train ML models to predict phenotypic class from cell features.
```
Comment (axiomcura) [inline 2.train_model/README.md:3]:
```suggestion
In this module, we trained ML models to predict phenotypic classes from cell features.
```

File: 2.train_model/README.md [hunk 2]
```diff
@@ -2,7 +2,10 @@
...
 
-We train the models in [train_model.ipynb](1.train_model.ipynb).
+In [train_multi_class_models.ipynb](train_multi_class_models.ipynb), we train models to predict which phenotypic class the cell features are from out of 15 possible classes (anaphase, metaphase, apoptosis, etc).
```
Comment (axiomcura) [inline 2.train_model/README.md]:
```suggestion
In [train_multi_class_models.ipynb](train_multi_class_models.ipynb), we are training models to predict the phenotypic class of the cell features from 15 possible classes (anaphase, metaphase, apoptosis, etc).
```
Feel free to reject suggestions if it does not convery your message.

File: 2.train_model/README.md [hunk 3]
```diff
@@ -2,7 +2,10 @@
...
+In [train_single_class_models.ipynb](train_single_class_models.ipynb), we train models to predict whether or not the cell features are from a particular phenotypic class or not.
```
Comment (axiomcura) [inline 2.train_model/README.md]:
```suggestion
In [train_single_class_models.ipynb](train_single_class_models.ipynb), we are training models to predict whether the cell features are from a particular phenotypic class or not.
```

File: 2.train_model/README.md (hunk 1 above)
Comment (roshankern) [reply 2.train_model/README.md:3]: I think the tense I used is correct here... maybe @gwaybio can confirm? If this isn't correct I will have to address it in the rest of the modules as well.

I remember when I settled for this tense and voice I was following the [Cell Health](https://github.com/broadinstitute/cell-health/tree/master/0.download-data) example. For example:
> In this module, we download all single-cell Cell Painting profiles generated for the Cell Health project.

File: 2.train_model/scripts/nbconverted/train_single_class_models.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+training_data = get_dataset(features_dataframe, data_split_indexes, "train")
+# training_data
+
+
+# In[3]:
+
+
+# specify model types and feature types
+model_types = ["final", "shuffled_baseline"]
+feature_types = ["CP", "DP", "CP_and_DP"]
```
Comment (axiomcura) [inline 2.train_model/scripts/nbconverted/train_single_class_models.py:54]: It seems that you are creating combinations with these two lists. Consider using `itertools.combinations` to remove the nested `for` loops and improve readability. 

There are other types of combinations [algorithms](https://docs.python.org/3/library/itertools.html#module-itertools) select that one that best suits for you. (very last table)

File: 2.train_model/scripts/nbconverted/train_single_class_models.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+print(f"Parameters being tested during grid search: {parameters}\n")
+
+# create grid search with cross validation with hypertuning params
+grid_search_cv = GridSearchCV(
+    log_reg_model, parameters, cv=straified_k_folds, n_jobs=-1, scoring="f1_weighted"
+)
+
+# train model on each combination of model type and feature type
+for model_type in model_types:
+    for feature_type in feature_types:
```
Comment (axiomcura) [inline 2.train_model/scripts/nbconverted/train_single_class_models.py]: Look at comment above.

File: 2.train_model/scripts/nbconverted/train_single_class_models.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+                training_data.loc[
+                    training_data["Mitocheck_Phenotypic_Class"] == phenotypic_class
+                ]
+            ).shape[0]
+            print(
+                f"Training {model_type} model on {feature_type} features for {phenotypic_class} phenotypic class with {phenotypic_class_counts} positive labels..."
+            )
+
+            # create deep copy of training data so we can make modifications without affecting original training data
+            class_training_data = training_data.copy(deep=True)
```
Comment (axiomcura) [inline 2.train_model/scripts/nbconverted/train_single_class_models.py]: Data duplication can be problematic if the dataset is large. Do you think adding a column to the dataframe would solve this issue or storing the edited labels into a list and then concatenating it to the original dataframe.

**Review by axiomcura — APPROVED** [review]
Nice PR :)! 

I left some comments and suggestions. Feel free to ask me any questions !

File: 2.train_model/README.md (hunk 1 above)
Comment (gwaybio) [reply 2.train_model/README.md:3]: Consistency is the most important. I tend to prefer present tense in analysis repositories, but past tense in papers.

File: 2.train_model/scripts/nbconverted/train_single_class_models.py (hunk 6 above)
Comment (roshankern) [reply 2.train_model/scripts/nbconverted/train_single_class_models.py]: I don't think it is worth worrying about the data duplication here as we know the training data dataframe will be relatively small and only one copy needs to be held in memory per for loop. This sample also allows use of the functions used for the training data in multi class models notebook (i.e. `get_X_y_data()`)

File: 2.train_model/README.md (hunk 2 above)
Comment (roshankern) [reply 2.train_model/README.md]: I kept my verb tense but included your other suggestions in https://github.com/WayScience/phenotypic_profiling_model/pull/25/commits/6038b55195d1339937b3f667eb9ed2400677c875

File: 2.train_model/README.md (hunk 3 above)
Comment (roshankern) [reply 2.train_model/README.md]: I kept my verb tense but included your other suggestion in https://github.com/WayScience/phenotypic_profiling_model/pull/25/commits/6038b55195d1339937b3f667eb9ed2400677c875

File: 2.train_model/scripts/nbconverted/train_single_class_models.py (hunk 4 above)
Comment (roshankern) [reply 2.train_model/scripts/nbconverted/train_single_class_models.py:54]: I ended up using `itertools.product`. Thanks for the suggestion!

**Pull request merged**